Rechargable spells

thokkat opened the pull request:
Most spells in G1 and rechargeable ones in G2 rely on `manaInvested` variable in script. This value is increased by one per iteration of `tickCast()` and passed to the corresponding script functions. The resulting npc mana stat reduction has to be done by engine for G1, G2  case is handled by script. `SPL_RECEIVEINVEST` for G1 and `SPL_STATUS_CANINVEST_NO_MANADEC` for G2  seem to be used as a waiting state if neither a new level is reached nor spell is cast.

If a spell's investing phase is interrupted by player interaction script function `Spell_ProcessMana_Release` is called to decide if spell should be cast or aborted.

In G1`S_MAGRUN` animation has no idle flag. I removed the check to let idle animation correctly play but don't know if this has other effects. There's is other G1 specific stuff like firerain and stormfist have `TARGET_COLLECT_ALL_FALLBACK_NONE` flag, but stormfist does no damage if flag is added to the `isSpellShoot()` check to make firerain effect play.

G1 also has no `Spell_Cast` script routine so only call it in G2 case.

Try reviewed (commented):
Posted a few questions about code, still reviewing

thokkat commented on game/world/objects/npc.cpp:
```diff
@@ -3447,26 +3457,35 @@ bool Npc::tickCast(uint64_t dt) {
 
-  if(castLvl<=15)
-    castLevel = CastState(castLevel+1);
+  ++manaInvested;
```
In G2 the amout of mana needed to reach a new cast level is constant per spell. But G1 script has lines like 
```
	if 	(manaInvested ==(SPL_SENDCAST_FIRESTORM*3)/10)
	||	(manaInvested ==(SPL_SENDCAST_FIRESTORM*5)/10)
	||	(manaInvested ==(SPL_SENDCAST_FIRESTORM*8)/10)
	{	return SPL_NEXTLEVEL;		};
```
where mana needed for next level can be arbitrary.

thokkat replied:
In G1 player's mana is reduced in 1-steps and script comments stated `manaInvested` is increased by 1 per `Spell_Logic` call by engine. There's no mana attribute handling in G1 scripts so I added `changeAttribute(ATR_MANA,-1,false);`

For G2 i tried to get the needed mana somehow from values like `const int STEP_ChargeFireball = 40;` but after that was unsatisfying I added a print command to scripts to check the value in vanilla. And indeed in G1 and G2 `manaInvested` is a `0,1,2,... ` sequence. But here mana attribute changes are handled by script.  

This handling in G2  inherits the vanilla bug that if e.g. players' mana is 70 and mana per level 50, spell is charged to level 2 and player's mana is set to 0. So one more level than mana would allow for if player's start mana is in spell's `[min,max)` interval.

Try replied:
In G2-notr, there is `Spell_Logic_SuckEnergy` only spell that uses `SPL_RECEIVEINVEST` code, do not decrement mana, seemingly relay on engine?

thokkat replied:
>Should it be 1 per SPL_RECEIVEINVEST ?

I retested, `
 SPL_RECEIVEINVEST;
  SPL_STATUS_CANINVEST_NO_MANADEC;
  SPL_NEXTLEVEL;
`  all give +1.

`SPL_DONTINVEST` is only used  in G2 to hold a maximal charged spell. `Spell_Logic` is not called anymore and spell is sent on player action. 

>In G2-notr, there is Spell_Logic_SuckEnergy only spell that uses SPL_RECEIVEINVEST code, do not decrement mana, seemingly relay on engine?

I thought this state was only in use by G1, missed that one. Mana decrement is handled in `Spell_Cast_SuckEnergy`.

In G1 there's line `if (manaInvested == 0) { return SPL_NEXTLEVEL;	};` in firerain file. Comment hints it's needed to play invest fire effect. Maybe it's something similar for `SuckEnergy`.

thokkat commented on game/world/objects/npc.cpp:
```diff
@@ -3479,10 +3500,17 @@ bool Npc::tickCast(uint64_t dt) {
   if(castLvl<0)
     return;
+  if(abort) {
+    const SpellCode code = SpellCode(owner.script().invokeManaRelease(*this,currentTarget,manaInvested));
```
>Apparently invokeManaRelease should be called when player releases action-button, and current logic is out of place.

It also needs to be accessed if mana reaches 0 in G1 because there's no mana check in script if total mana is enough to cast spell. I look for a cleaner solution.

thokkat commented:
In G2 ai doesn't use rechargeable spells. In G1 `mana` parameter in `ai_readyspell` is used to define max mana to be used. For non rechargeable spells `mana` should equal needed mana amount  for spell. In other case it's based on distance to target and available mana.
```
			else if ( (Npc_GetDistToNpc(slf,oth)>(SPL_RANGE_WINDFIST/2)) && (slf.attribute[ATR_MANA]>=SPL_SENDCAST_WINDFIST/2) )
			{
				PrintDebugNpc	(PD_ZS_CHECK,	"...mittlerer!");
				B_DrawSpell		(slf,	SPL_WINDFIST,	SPL_SENDCAST_WINDFIST/2);
```

In `castSpell` routine spells by ai are made immediately cast and bypassing `tickCast`. Not a problem for G2 because all spells used by ai get immediate `SPL_SENDCAST` spellcode in `beginCastSpell` anyway. But in G1 spells have this one by one mana stat reduction and `tickCast` is needed to charge to correct level.

Try commented on game/world/objects/npc.h:
```diff
@@ -550,6 +551,8 @@ class Npc final {
     // spell cast
     CastState                      castLevel        = CS_NoCast;
+    int32_t                        manaInvested     = 0;
+    int32_t                        spellMana        = 1;
```
`aiExpectedInvest` maybe? to make use-case more clear?

Try reviewed (commented):
Also FYI: few changes to cast controls e64325f5

thokkat commented on game/world/objects/npc.cpp:
```diff
@@ -3486,6 +3513,16 @@ void Npc::endCastSpell() {
   castLevel = CastState(castLvl+CS_Cast_0);
   }
 
+void Npc::releaseSpell() {
```
>What place this function takes in control flow?`

Just misunderstood what you meant with `current logic is out of place`.

thokkat commented:
For G1 check for zero mana is now done for player only and an assert is added for npc's. Assert should never be triggered because script cheats mana for npcs. In case a mod chages this.

I modified `playercontrol` to have spells charging as long action key (and forward in g1 controls case) is pressed.

>In G1S_MAGRUN animation has no idle flag. I removed the check to let idle animation correctly play but don't know if this has other effects

Reverted the idle check remove because bow/cbow loading animation is not working correctly.

Try commented on game/game/serialize.h:
```diff
@@ -33,7 +33,7 @@ class SaveGameHeader;
   public:
     enum Version : uint16_t {
-      Current = 43
+      Current = 44
```
note: save file version was bumped in 57241ed3

Try reviewed (commented):
Looks good, can merge once conflicts are resolved

thokkat commented on game/world/objects/npc.cpp:
```diff
@@ -3481,11 +3506,18 @@ bool Npc::tickCast(uint64_t dt) {
   }
 
-void Npc::endCastSpell() {
+void Npc::endCastSpell(bool playerCtrl) {
```
Need to know if spell was ended by player interruption or got send signal from script. For example in G1 if player has 15 mana and cast firerain with 15 mana it's enough to cast spell. But if `invokeMana` would be always called if npc is player spell would be aborted.

Try replied:
I've tested current state of path, with G1(Steam):
fire-rain doesn't emit any effect, seem to be broken. Other spell are OK.

thokkat replied:
yeah there's a problem with spell detection
>firerain and stormfist have TARGET_COLLECT_ALL_FALLBACK_NONE flag, but stormfist does no damage if flag is added to the `isSpellShoot()` check to make firerain effect play.

Firerain invest effect is also missing. Same goes for G1 maintenance spells like pyrokinesis where invest phase and cast phase are identical. I only tested that firerain invest effect can be played but a clean implementation is out of scope for me so i didn't touch any of those.

Try commented on game/world/objects/npc.cpp:
```diff
@@ -3448,26 +3461,38 @@ bool Npc::tickCast(uint64_t dt) {
+    case SpellCode::SPL_RECEIVEINVEST:
+    case SpellCode::SPL_STATUS_CANINVEST_NO_MANADEC: {
+      auto& spl = owner.script().spellDesc(active->spellId());
       visual.setMagicWeaponKey(owner,SpellFxKey::Invest,castLvl+1);
```
Tested today:
apparently `setMagicWeaponKey` must go into `SPL_NEXTLEVEL`, otherwise any spell in G2 has broken sound/visual

Try commented:
Merged, thanks!
